File: ssh_monitor.py
"""
SSH Log Monitor - Canlı log izleme modülü
Uzak sunucuya SSH ile bağlanıp log dosyasını gerçek zamanlı izler.
"""

import paramiko
import threading
import queue
import re
from datetime import datetime
from typing import Generator, Dict, Any, Optional
import time


class SSHLogMonitor:
    """SSH üzerinden log dosyası izleme sınıfı"""
    
    # Log satırı parsing pattern'leri
    THREAT_PATTERNS = {
        'BRUTE_FORCE': {
            'pattern': r'Failed password|Failed publickey|authentication failure.*rhost=',
            'severity': 'HIGH',
            'description': 'Başarısız giriş denemesi tespit edildi'
        },
        'INVALID_USER': {
            'pattern': r'Invalid user|user unknown',
            'severity': 'HIGH', 
            'description': 'Geçersiz kullanıcı ile giriş denemesi'
        },
        'ROOT_LOGIN': {
            'pattern': r'Accepted.*for root|session opened.*root',
            'severity': 'CRITICAL',
            'description': 'Root kullanıcısı ile giriş yapıldı'
        },
        'LOGIN_SUCCESS': {
            'pattern': r'Accepted password|Accepted publickey',
            'severity': 'INFO',
            'description': 'Başarılı giriş'
        },
        'SESSION_OPENED': {
            'pattern': r'session opened|New session',
            'severity': 'INFO',
            'description': 'Oturum açıldı'
        },
        'SESSION_CLOSED': {
            'pattern': r'session closed|Removed session|logged out',
            'severity': 'INFO',
            'description': 'Oturum kapatıldı'
        },
        'SUDO_USAGE': {
            'pattern': r'sudo:.*COMMAND=',
            'severity': 'MEDIUM',
            'description': 'Sudo komutu çalıştırıldı'
        },
        'CONNECTION_CLOSED': {
            'pattern': r'Connection closed|Disconnected from',
            'severity': 'INFO',
            'description': 'Bağlantı kapatıldı'
        }
    }
    
    def __init__(self):
        self.client: Optional[paramiko.SSHClient] = None
        self.channel = None
        self.is_connected = False
        self.is_streaming = False
        self.log_queue = queue.Queue()
        self.stream_thread = None
        self.connection_info = {}
        
    def connect(self, host: str, username: str, password: str = None, 
                key_path: str = None, port: int = 22) -> Dict[str, Any]:
        """SSH bağlantısı kur"""
        try:
            self.client = paramiko.SSHClient()
            self.client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
            
            connect_kwargs = {
                'hostname': host,
                'port': port,
                'username': username,
                'timeout': 10
            }
            
            if key_path:
                connect_kwargs['key_filename'] = key_path
            elif password:
                connect_kwargs['password'] = password
            else:
                return {'success': False, 'error': 'Şifre veya SSH anahtarı gerekli'}
            
            self.client.connect(**connect_kwargs)
            self.is_connected = True
            self.connection_info = {
                'host': host,
                'port': port,
                'username': username,
                'connected_at': datetime.now().isoformat()
            }
            
            return {
                'success': True,
                'message': f'{username}@{host}:{port} bağlantısı kuruldu',
                'connection': self.connection_info
            }
            
        except paramiko.AuthenticationException:
            return {'success': False, 'error': 'Kimlik doğrulama başarısız - şifre veya anahtar hatalı'}
        except paramiko.SSHException as e:
            return {'success': False, 'error': f'SSH hatası: {str(e)}'}
        except Exception as e:
            return {'success': False, 'error': f'Bağlantı hatası: {str(e)}'}
    
    def disconnect(self) -> Dict[str, Any]:
        """SSH bağlantısını kapat"""
        self.is_streaming = False
        
        if self.channel:
            try:
                self.channel.close()
            except:
                pass
            self.channel = None
            
        if self.client:
            try:
                self.client.close()
            except:
                pass
            self.client = None
            
        self.is_connected = False
        return {'success': True, 'message': 'Bağlantı kapatıldı'}
    
    def parse_log_line(self, line: str) -> Dict[str, Any]:
        """Log satırını parse et ve tehdit analizi yap"""
        # Temel log bilgilerini çıkar
        parsed = {
            'raw': line.strip(),
            'timestamp': datetime.now().isoformat(),
            'threat_type': None,
            'severity': 'INFO',
            'description': line.strip(),
            'source_ip': None,
            'username': None
        }
        
        # Timestamp pattern (syslog formatı)
        timestamp_match = re.search(
            r'(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}[^\s]*|\w{3}\s+\d+\s+\d{2}:\d{2}:\d{2})',
            line
        )
        if timestamp_match:
            parsed['log_timestamp'] = timestamp_match.group(1)
        
        # IP adresi çıkar
        ip_match = re.search(r'from\s+(\d+\.\d+\.\d+\.\d+)|rhost=(\d+\.\d+\.\d+\.\d+)', line)
        if ip_match:
            parsed['source_ip'] = ip_match.group(1) or ip_match.group(2)
        
        # Kullanıcı adı çıkar
        user_match = re.search(r'for\s+(\w+)|user[=\s]+(\w+)|USER=(\w+)', line)
        if user_match:
            parsed['username'] = user_match.group(1) or user_match.group(2) or user_match.group(3)
        
        # Tehdit tespiti
        for threat_type, config in self.THREAT_PATTERNS.items():
            if re.search(config['pattern'], line, re.IGNORECASE):
                parsed['threat_type'] = threat_type
                parsed['severity'] = config['severity']
                parsed['description'] = config['description']
                
                # Detaylı açıklama oluştur
                if parsed['source_ip']:
                    parsed['description'] += f" (Kaynak: {parsed['source_ip']})"
                if parsed['username']:
                    parsed['description'] += f" - Kullanıcı: {parsed['username']}"
                break
        
        return parsed
    
    def start_log_stream(self, log_path: str = '/var/log/auth.log') -> Generator[Dict[str, Any], None, None]:
        """Log dosyasını izlemeye başla ve satırları yield et"""
        if not self.is_connected or not self.client:
            yield {'error': 'SSH bağlantısı yok', 'success': False}
            return
        
        self.is_streaming = True
        
        try:
            # tail -n 50 -f komutu ile son 50 satırı al ve takibe devam et
            stdin, stdout, stderr = self.client.exec_command(
                f'tail -n 50 -f {log_path}',
                get_pty=True
            )
            
            self.channel = stdout.channel
            self.channel.settimeout(0.5)  # Non-blocking read
            
            buffer = ""
            
            while self.is_streaming:
                try:
                    if self.channel.recv_ready():
                        data = self.channel.recv(4096).decode('utf-8', errors='ignore')
                        buffer += data
                        
                        # Satır satır işle
                        while '\n' in buffer:
                            line, buffer = buffer.split('\n', 1)
                            if line.strip():
                                parsed = self.parse_log_line(line)
                                yield parsed
                    else:
                        time.sleep(0.1)  # CPU kullanımını azalt
                        
                except socket.timeout:
                    continue
                except Exception as e:
                    if self.is_streaming:
                        yield {'error': str(e), 'success': False}
                    break
                    
        except Exception as e:
            yield {'error': f'Stream hatası: {str(e)}', 'success': False}
        finally:
            self.is_streaming = False


# Global instance
ssh_monitor = SSHLogMonitor()


def get_monitor() -> SSHLogMonitor:
    """Global SSH monitor instance'ını döndür"""
    return ssh_monitor

File: LogIz-LogIz/scripts/ssh-monitor.js
// scripts/ssh-monitor.js
const { Client } = require('ssh2');

const SSH_CONFIG = {
    host: process.env.SSH_HOST || 'localhost',
    port: parseInt(process.env.SSH_PORT || '22'),
    username: process.env.SSH_USER || 'root',
    password: process.env.SSH_PASSWORD,
    privateKey: process.env.SSH_KEY_PATH ? require('fs').readFileSync(process.env.SSH_KEY_PATH) : undefined,
};

const LOG_API_URL = process.env.LOG_API_URL || 'http://localhost:3000/api/live-stream';

// Threat Patterns (Partial list ported from Python)
const PATTERNS = [
    { name: 'BRUTE_FORCE', regex: /Failed password|Failed publickey|authentication failure.*rhost=/, severity: 'HIGH' },
    { name: 'ROOT_LOGIN', regex: /Accepted.*for root|session opened.*root/, severity: 'CRITICAL' },
    { name: 'LOGIN_SUCCESS', regex: /Accepted password|Accepted publickey/, severity: 'INFO' },
    { name: 'SESSION_OPENED', regex: /session opened|New session/, severity: 'INFO' },
    { name: 'INVALID_USER', regex: /Invalid user|user unknown/, severity: 'HIGH' },
];

function startMonitor() {
    const conn = new Client();

    conn.on('ready', () => {
        console.log('✅ SSH Connection established');

        conn.exec('tail -f /var/log/auth.log', (err, stream) => {
            if (err) throw err;

            stream.on('close', (code, signal) => {
                console.log('Stream :: close :: code: ' + code + ', signal: ' + signal);
                conn.end();
            }).on('data', (data) => {
                const lines = data.toString().split('\n');
                lines.forEach(line => {
                    if (!line.trim()) return;

                    // Basic Analysis
                    let severity = 'INFO';
                    let threat = null;

                    for (const p of PATTERNS) {
                        if (p.regex.test(line)) {
                            severity = p.severity;
                            threat = p.name;
                            break;
                        }
                    }

                    // Extract IP
                    const ipMatch = line.match(/from\s+(\d+\.\d+\.\d+\.\d+)|rhost=(\d+\.\d+\.\d+\.\d+)/);
                    const ip = ipMatch ? (ipMatch[1] || ipMatch[2]) : 'Unknown';

                    // Send to API
                    const payload = {
                        message: line.trim(),
                        source: 'ssh-monitor',
                        ip: ip,
                        severity: severity,
                        threat: threat
                    };

                    fetch(LOG_API_URL, {
                        method: 'POST',
                        headers: { 'Content-Type': 'application/json' },
                        body: JSON.stringify(payload)
                    }).catch(e => console.error('API Error:', e.message));

                    console.log(`[${severity}] ${line.substring(0, 50)}...`);
                });
            }).stderr.on('data', (data) => {
                console.log('STDERR: ' + data);
            });
        });
    }).on('error', (err) => {
        console.error('SSH Connection Error:', err.message);
        setTimeout(startMonitor, 5000); // Reconnect logic
    }).connect(SSH_CONFIG);
}

if (require.main === module) {
    if (!SSH_CONFIG.username || (!SSH_CONFIG.password && !SSH_CONFIG.privateKey)) {
        console.error('❌ Missing SSH credentials in env variables (SSH_USER, SSH_PASSWORD/SSH_KEY_PATH)');
        process.exit(1);
    }
    console.log('🚀 Starting SSH Monitor...');
    startMonitor();
}
